### DISCUSSION TITLE
Building Modern UI

### DISCUSSION THREAD

**Discussion by gabriellanata** [MobileNativeFoundation/discussions#53] [General]
Hello!

I'm curious to hear what other companies are doing to improve the developer experience of working with UI. 

SwiftUI and Jetpack Compose showed us what the future of building UI looks like for Apple and Google, which have both embraced a declarative and reactive approach. But having limited OS support, lacking some initial functionality and sometimes being very buggy has prevented many from adopting them. 

Has anyone integrated SwiftUI or Jetpack Compose in a large codebase? How has it worked out? If not, are you still building UI using standard native code or have you created/adopted a custom framework?

**Comment by gabriellanata** [comment]
I'll start by sharing what we've done at Lyft on iOS.

We have a very large and mature design system which wraps UIKit and provides all sorts of components. The API for using these components was originally designed to feel very "native", as we strongly believed that this lowered the learning curve by leveraging developer's existing familiarity with UIKit. However, we slowly ended up adding a bunch of conveniences methods and wrapping views to deal with some annoyances when writing autolayout.

When SwiftUI was announced we loved the simplicity and readability it brought. But it was many years away from us being able to adopt it because we still support iOS 11, not including the bugs and architecture compatibility issues. A little over a year ago I started working on modernizing our UI libraries to feel more like SwiftUI, the idea was that this would prepare us for a possible transition in the future. This grew into a large project and we now have a very mature Declarative UI library that makes UIKit use SwiftUI style syntax and reactiveness, but still feels very much like UIKit. The responses have been great and it has definitely improved the experience of working with UI for us. The most exiting part is that you can actually code review what a screen looks like without seeing a screenshot or running the app, simply because the code is compact and readable.

I was planning on open sourcing this as part of the Mobile Native Foundation, so I want to see if this is a similar experience others have had. Let me know!

**Reply by shahzadmajeed** [reply]
@gabriellanata did you end-up open sourcing UI building framework you talked about above? I'm interested in seeing how you guys did it.

**Comment by akbashev** [comment]
Won't say out app is very large at the moment (Holidu), but in the end I've come up with same solution actually.

Our iOS app is 100% Swift and basically use `UIKit` components with different extensions on it. `UICollectionView` used a lot, so we've come up with internal solution around `UICollectionView` and `UICollectionViewDataSource`—it allows us to easily split UI into multiple sections inside one collection view. It gave us also an option to easily share sections between different screens. Apple then introduced `UICollectionViewDiffableDataSource` with kinda same approach, so think in we were on right way.

As already mentioned, when SwiftUI was announced it was very promising UI framework, but reality is—we also need to support iOS 11 (even iOS 10 at that days). Also first version of framework had some issues, so would say it was not even considered at the beginning. Things changed after SwiftUI 2.0 introduction, cause some parts of ecosystem like WatchOS only supports SwiftUI now. Android is also moving towards this way with Jetpack Compose. Seems like in several years declarative UI will be everywhere.

We've started to think about it, and also came up with some solutions to prepare us for transition, and decision was to move forwards in two steps:

1. I've wrote simple API around `UIStackView` to mimic `SwiftUI` and called it `KindaDeclarativeUI™️` 😅 It's a bit hacky and `UIStackView` itself is not very performant, but It gives us opportunity to write code like this:
```
    StackScrollView {
      VerticalStack(alignment: .leading, spacing: 16) {
        VerticalStack(alignment: .leading, spacing: 6) {
          self.emailField.frame(width: .infinity, height: 44)
          self.emailFieldErrorLabel
          FlatStack(alignment: .trailing) {
            self.passwordField.frame(width: .infinity, height: 44)
            self.showButton
          }
          self.passwordFieldErrorLabel
        }
        self.loginButton.frame(width: .infinity, height: 48)
        VerticalStack(alignment: .leading, spacing: 12) {
          self.recoverButton
          self.signUpButton.frame(width: .infinity)
        }
      }.padding(top: 16, left: 16, bottom: 16, right: 16)
    }.reference(to: &self.contentView).add(to: self)
```
This approach gave us readability and saved a lot of time already. All alignments are same as in `SwiftUI` and it centred by default. As you can see you also can modify views same way, like `.padding()` or `.frame`. Also `UIStackView` gives some advantages like ability to hide view, so that stack view will relayout itself, e.g. `self.emailFieldErrorLabel` in code is only visible when there are some errors. As you know `UIStackView` is bad performance wise, so we're not using it for complex screens. In ideal world would say it's better to write some layouting using frames and wrap in kinda same API. But in real world this solutions was written in a week and would say it's saved us this week afterwards. It's not open sourced, but if someone's interested—please write.

2. Next step is just to start using `SwiftUI`—iOS 11 usage is very low, iOS 12 usage is getting lower also. Future is coming. 😅 
But there are some problems as we can't just move towards it without changing architecture a bit. Also I'm not a big fan using `ViewModel` with `SwiftUI`, so after some research we've decided to move forwards Redux kinda architecture with states. Long story short I've chosen [Composable Architecture](https://github.com/pointfreeco/swift-composable-architecture) as a base architecture for our transition, but would say it's not a topic of this thread.

So next month I'll try to rewrite some simple screen using `SwiftUI` and `TCA`, show it to iOS 13+ users, add some metrics and make some conclusions afterwards. Not sure about metrics though, cause think readability and sustainability is more important here. Also want to show this to our frontend guys, cause they are using Redux and would be interesting to see if they will understand something. 🙂

Would be interesting to hear if someone went same way (UI and architecture wise), what was a motivation and results in the end. Thanks!

**Reply by gabriellanata** [reply]
Yeah our implementation looks very similar. Seems like we're on the same path which helps validate my theory that at least a few companies could use what we've built. Thanks for sharing!

**Reply by stuartjmoore** [reply]
I’ve been playing with the same `KindaDeclarativeUI™️` idea since just before SwiftUI was announced. It’s gone through many stages (using data sources, creating a “stack view controller”, using function/result builders, to just being a convenience on stack view). Nothing seemed revolutionary enough that I kept second guessing it.

Now I’m thinking it’s some sort of template engine that allows us to reuse the same layouts we get from design without coupling them into one massive view controller.

**Comment by erichoracek** [comment]
Thanks for starting this thread! I can chime in on how we work with UI at Airbnb:

We just open sourced and [published a blog post with details](https://medium.com/airbnb-engineering/introducing-epoxy-for-ios-6bf062be1670) on our declarative UI systems. We use a library we've built over the last few years and recently open sourced named [Epoxy](https://github.com/airbnb/epoxy-ios) to enable declarative management of both the view and view controller hierarchy.

We've found that using a declarative approach for our UI has made also it possible to adopt Unidirectional Data Flow for our feature architecture—we use an approach with some similarities to TCA.

We're super excited about SwiftUI, but have found that it's not a good fit for us in the [short to medium term](https://github.com/airbnb/epoxy-ios/wiki/FAQ#why-would-i-use-epoxy-and-uikit-instead-of-swiftui).

**Comment by thedrick** [comment]
Hey everyone, excited to hear more from other folks about how they are building UI! We actually [just open sourced](https://github.com/airbnb/epoxy-ios/pull/27) our version of declarative UI for building components called `LayoutGroups` that we've put inside of `Epoxy` that @erichoracek mentioned above. It's quite similar to what @akbashev and @gabriellanata mentioned, though we rolled our own layout containers instead of using `UIStackView`. I'd love for folks to check it out and let us know what you think!
